File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        product = 1
        count = 0
        for num in nums:
            if num == 0:
                count += 1
                continue
            product *= num

        
        if count > 1:
            return [0] * len(nums)

        res = []

        for num in nums:
            if count > 0 and num == 0:
                res.append(product)
            elif count > 0:
                res.append(0)
            elif count == 0:
                res.append(product // num)
        return res

File: Data Structures & Algorithms/minimum-number-of-operations-to-make-array-empty/submission-2.cpp
class Solution {
public:
    int minOperations(vector<int>& nums) {
        unordered_map<int, int> nums_count;
        for (int &x : nums) {
            nums_count[x] += 1;
        }
        int res = 0;
        for (auto &[key, value] : nums_count) {
            if (value == 1)
                return -1;

            if (value % 3 == 0)
                res += value / 3;
            
            if (value % 3 == 1)
                res += (value / 3) - 1 + 2;
            
            if (value % 3 == 2)
                res += (value / 3) + 1;

        }

        return res;
    }
};